File: src/main/java/com/distributed/lab7/Matrix.java
package com.distributed.lab7;

import java.util.stream.IntStream;

public class Matrix {
    public static double[][] generateMatrix(int size, int maxElem) {
        double[][] matrix = new double[size][];
        for (int i = 0; i < matrix.length; i++) {
            matrix[i] = new double[size];
            for (int j = 0; j < matrix[i].length; j++) {
                matrix[i][j] = Math.random() * maxElem;
            }
        }
        return matrix;
    }

    public static boolean check(double[][] left, double[][] right) {
        for (int i = 0; i < left.length; i++) {
            for (int j = 0; j < left[i].length; j++) {
                if (left[i][j] != right[i][j])
                    return false;
            }
        }
        return true;
    }

    public static void multiplication(double[][] left, double[][] right, double[][] nativeResult) {
        for (int i = 0; i < left.length; i++) {
            for (int j = 0; j < left[i].length; j++) {
                for (int k = 0; k < left[i].length; k++) {
                    nativeResult[i][j] += left[i][k] * right[k][j];
                }
            }
        }
    }

    public static void parallelMultiplication(double[][] left, double[][] right, double[][] parallelResult) {
        IntStream.range(0, left.length).parallel().forEach(i -> {
            for (int j = 0; j < left.length; j++) {
                for (int k = 0; k < left.length; k++) {
                    parallelResult[i][j] += left[i][k] * right[k][j];
                }
            }
        });
    }
}

File: module_1/src/main/csharp/Lab7/Matrix.cs
using System;
using System.Threading.Tasks;

namespace Lab7
{
    public static class Matrix
    {
        public static double[][] GenerateMatrix(int size, int maxElem)
        {
            var matrix = new double[size][];
            for (var i = 0; i < matrix.Length; i++)
            {
                matrix[i] = new double[size];
                for (var j = 0; j < matrix[i].Length; j++)
                    matrix[i][j] = new Random().NextDouble() * maxElem;
            }
            return matrix;
        }

        public static bool Check(double[][] left, double[][] right)
        {
            for (var i = 0; i < left.Length; i++)
            for (var j = 0; j < left[i].Length; j++)
                if (Math.Abs(left[i][j] - right[i][j]) > 1e-5)
                    return false;
            return true;
        }
        
        
        public static void Multiplication(double[][] left, double[][] right, double[][] result)
        {
            for (var i = 0; i < left.Length; i++)
            for (var j = 0; j < left[i].Length; j++)
            for (var k = 0; k < left[i].Length; k++) 
                result[i][j] += left[i][k] * right[k][j];
        }
        
        public static void ParallelMultiplication(double[][] left, double[][] right, double[][] result)
        {
            Parallel.For(0, left.Length, i =>
            {
                for (var j = 0; j < left[i].Length; j++)
                {
                    for (var k = 0; k < left[i].Length; k++)
                        result[i][j] += left[i][k] * right[k][j];
                }
            });
        }
    }
}